### PR TITLE
Preserve segmentation mask dtype and class indices in preprocessing layers

### PR DESCRIPTION
**Pull request by rstar327**
## Description

Image preprocessing layers corrupt `segmentation_masks` in two ways when masks
are passed via the dict call (`layer({"images": ..., "segmentation_masks": ...})`):

1. **Invented class values.** Layers that resize masks use bilinear
   interpolation, which produces fractional labels between real classes (e.g.
   a mask with classes `2` and `4` gains a `3` that never existed).
2. **dtype loss.** `transform_segmentation_masks` delegates to
   `transform_images`, which casts to the layer's `compute_dtype` (usually
   `float32`), so a `uint8` class-index mask comes back as `float32`.

Related to #20857. This takes the per-layer approach: each affected class is
fixed so masks are never interpolated/blended into new values and keep their
original integer dtype — rather than casting the result back at the end (which
would silently round invented labels like `3.0` into a bogus class `3`).

### Fixes

- **`CenterCrop`** / **`RandomCrop`** — the resize fallback (used when the
  target is larger than the input) now uses `interpolation="nearest"` for
  masks, and the original dtype is restored. Nearest sampling reuses exact
  source pixels, so no new classes appear.
- **`RandomFlip`** — flipping only reorders pixels; masks are now flipped in
  their original dtype without the `float32` cast.
- **`CutMix`** — pasting a region between samples is an exact copy; masks are
  pasted in their original dtype.

The image / label / bounding-box paths are unchanged (images still use
bilinear).

### Out of scope

`MixUp` and `AugMix` *blend* samples (`w * a + (1 - w) * b`), which is
ill-defined for discrete class-index masks and currently errors on integer
masks. The correct behavior there (leave masks untouched vs. require one-hot)
is a separate design decision and is intentionally not addressed here.

### Tests

Added a regression test to each of the four layers asserting that a `uint8`
mask with classes `{0, 4}` keeps its dtype and gains no new class values
(`CenterCrop`/`RandomCrop` tests exercise the resize path). All four layers'
existing suites pass on the TensorFlow and NumPy backends.

## Contributor Agreement

**Please review our [AI-Assisted Contribution Policy](../CONTRIBUTING.md#ai-assisted-contribution-policy) and check all boxes below before submitting your PR for review:**

- [x] I am a human, and not a bot.
- [x] I will be responsible for responding to review comments in a timely manner.
- [x] I will work with the maintainers to push this PR forward until submission.

### CODE REVIEW

**Comment by rstar327** [comment]
Fixed by forwarding `training` to `transform_images`, and added the `training=False` no-op assertions to the test. Verified the test fails without the fix and passes with it.

File: keras/src/layers/preprocessing/image_preprocessing/center_crop.py [hunk 1]
```diff
@@ -177,11 +177,18 @@ def _get_clipped_bbox(bounding_boxes, h_end, h_start, w_end, w_start):
...
+        masks = self.backend.convert_to_tensor(segmentation_masks)
+        in_dtype = masks.dtype
+        outputs = self._center_crop(masks, interpolation="nearest")
+        return self.backend.cast(outputs, in_dtype)
 
     def transform_images(self, images, transformation=None, training=True):
+        return self._center_crop(images, interpolation="bilinear")
+
+    def _center_crop(self, images, interpolation="bilinear"):
         inputs = self.backend.cast(images, self.compute_dtype)
```
Comment (hertschuh) [inline keras/src/layers/preprocessing/image_preprocessing/center_crop.py]: `inputs = self.backend.cast(images, self.compute_dtype)`

This line should be moved to `transform_images`. We don't want it for masks.

File: keras/src/layers/preprocessing/image_preprocessing/center_crop.py [hunk 2]
```diff
@@ -177,11 +177,18 @@ def _get_clipped_bbox(bounding_boxes, h_end, h_start, w_end, w_start):
...
-        return self.transform_images(
-            segmentation_masks, transformation, training=training
-        )
+        # Segmentation masks hold discrete class indices. Use nearest-neighbor
+        # interpolation on the resize path so no new class values are invented,
+        # and restore the original (typically integer) dtype.
+        masks = self.backend.convert_to_tensor(segmentation_masks)
+        in_dtype = masks.dtype
+        outputs = self._center_crop(masks, interpolation="nearest")
+        return self.backend.cast(outputs, in_dtype)
```
Comment (hertschuh) [inline keras/src/layers/preprocessing/image_preprocessing/center_crop.py]: Remove the cast once my other comment below is addressed.

File: keras/src/layers/preprocessing/image_preprocessing/random_crop.py [hunk 3]
```diff
@@ -128,7 +128,9 @@ def get_random_transformation(self, data, training=True, seed=None):
 
         return h_start, w_start
 
-    def transform_images(self, images, transformation, training=True):
+    def transform_images(
+        self, images, transformation, training=True, interpolation="bilinear"
+    ):
```
Comment (hertschuh) [inline keras/src/layers/preprocessing/image_preprocessing/random_crop.py]: Do not change this signature.

Also, the issue is that this method casts inputs.

Instead, use the same pattern as CenterCrop:
- create a utility private method `_center_crop` that takes an interpolation argument
- move most of the code here to `_center_crop` and here call `_center_crop`
- but keep the `self.backend.cast(images, self.compute_dtype)` here in `transform_images`
- `transform_segmentation_masks` calls `_center_crop` but no cast is needed

**Review by hertschuh — CHANGES_REQUESTED** [review]
Thanks for addressing this. A few comments:

**Comment by rstar327** [comment]
Thanks! Addressed all three: moved the `compute_dtype` cast into `transform_images`, added private `_center_crop`/`_random_crop` helpers (with the `interpolation` arg) so `transform_segmentation_masks` crops without casting, and kept the original `transform_images` signature. One note: I left an images-only cast after the resize fallback in `RandomCrop.transform_images` because TF upcasts float16→float32 there; masks are unaffected.

**Review by hertschuh — APPROVED** [review]
Thanks for fixing this!
